File: cpp/0040_combinationSum2.cpp
// 与0039的不同在于每个元素只能选择一次
// 回溯范围：遍历所有的组合
// 剪枝条件：当目前的sum已经大于target（需要先排序);当目前的元素和上一元素重复
// 终止：sum==target
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> res; 
        vector<int> comb;
        sort(candidates.begin(),candidates.end());
        backtrack(res,candidates,target,comb,0);
        return res;
    }
    void backtrack(vector<vector<int>> &res, vector<int>& candidates, int target, vector<int>&comb, int idx)
    {
        if(target==0)
        {
            res.push_back(comb);
            return;
        }
        for(int i=idx;i<candidates.size();i++)
        {
            if(target-candidates[i]<0)
                return;
            if(i>idx && candidates[i]==candidates[i-1])
                continue;

            comb.push_back(candidates[i]);
            backtrack(res,candidates,target-candidates[i],comb,i+1);
            comb.pop_back();
        }
    }
};

File: java/0040_combinationSum2.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> comb = new ArrayList<Integer>();
        Arrays.sort(candidates);
        backtrack(candidates,res,comb,target,0);
        return res;
    }
    public void backtrack(int[] candidates,List<List<Integer>> res,List<Integer> comb,int target,int pos)
    {
        if(target==0)
        {
            res.add(new ArrayList<Integer>(comb));
            return;
        }
        for(var i=pos;i<candidates.length;i++)
        {
            if(target-candidates[i]<0)
                return;
            if(i>pos && candidates[i]==candidates[i-1])
                continue;
            comb.add(candidates[i]);
            backtrack(candidates,res,comb,target-candidates[i],i+1);
            comb.remove(comb.size()-1);
        }
    }
}
